### PR TITLE
class to limit the output to n lines + class to manage time measures

### PR DESCRIPTION
**Pull request by ESeNonFossiIo**

### CODE REVIEW

File: include/utilities.h [hunk 1]
```diff
@@ -79,6 +79,27 @@ bool copy_file(const std::string &files, const std::string &destination);
...
+ * This class rewrite @p n_max lines of output
+ */
+class fixed_lines
+{
+public:
+  fixed_lines(int, std::ostream &stream_out = std::cout);
+  int get_current_line();
+  void goto_previous_line(int n_line = 1, bool erase=false);
+  void print_line(std::string &txt, bool erase=true);
+
```
Comment (ESeNonFossiIo) [inline include/utilities.h:334]: I would like to overload the operator << but I got a lot of troubles doing it...

File: source/utilities.cc [hunk 2]
```diff
@@ -150,3 +150,48 @@ bool rename_file(const std::string &file, const std::string &new_file)
...
+}
+
+void
+fixed_lines::goto_previous_line(int n_line, bool erase)
+{
+  if (curr_line>0)
+    for (unsigned int i = 0; i<n_line; ++i )
+      {
+        // go to the previous line:
+        std::cout << "\e[A";
```
Comment (ESeNonFossiIo) [inline source/utilities.cc]: this should be:

``` bash
stream_out << "\e[A";
```

The compiler returns 

``` bash
error: invalid operands to binary expression
```

**Comment by luca-heltai** [comment]
This works for me:

``` C++
// endl example
#include <iostream>     // std::cout, std::end

template<typename Stream>
class FilteredStream {
public:
  FilteredStream(Stream &out, unsigned int n) :
    n_lines(n),
    out(out)
  {};

  template<typename OBJ>
  FilteredStream<Stream>& operator<<(OBJ &o) {
    out << o;
    return *this;
  };

  FilteredStream<Stream>& operator<< (std::ostream& (*p) (std::ostream &)) {
    out << p;
    return *this;
  }

  template <typename S, typename T> friend FilteredStream<S>
  &operator << (FilteredStream<S> &, const T &);

  Stream & get_stream() {
    return out;
  };

private:
  unsigned int n_lines;
  Stream &out;
};

template <typename S, typename T>
inline
FilteredStream<S> &operator<< (FilteredStream<S> &log, const T &t)
{
  log.get_stream() << t;
  return log;
}


int main () {

  int a=100;
  double b=3.14;

  FilteredStream<std::ostream> out(std::cout, 1);

  out << a;
  out << std::endl;              // manipulator inserted alone
  out << b << std::endl << a*b << std::endl;  // manipulator in concatenated insertion

  return 0;
}
```

File: include/utilities.h (hunk 1 above)
Comment (luca-heltai) [inline include/utilities.h:334]: Take a look at 

`deal.II/base/logstream.h`

which is the one that creates deallog. they do something special whenever a new endl is encountered.

File: include/utilities.h [hunk 3]
```diff
@@ -79,6 +121,201 @@ bool copy_file(const std::string &files, const std::string &destination);
...
+ * - get_shell_cols returns the number of coloumns of the shell.
+ * - get_shell_rows returns the number of rows of the shell.
+ * - get_current_line returns the current line the class is writing.
+ *
+ */
+template<typename Stream>
+class FilteredStream
+{
+public:
+  FilteredStream( Stream &stream_out = std::cout,
```
Comment (luca-heltai) [inline include/utilities.h]: if you put the default here, you have to put the default also in the `template`

File: include/utilities.h [hunk 4]
```diff
@@ -31,6 +32,47 @@ void smart_delete (SmartPointer<TYPE> &sp) DEAL_II_DEPRECATED;
 std::string demangle(const char *name);
 
 /**
+ * This function collects some time utilities:
+ *  - sleep(unsigned int t) freezes the thread for t milliseconds;
```
Comment (luca-heltai) [inline include/utilities.h]: Moove the documentation of the members on top of each member...

File: include/utilities.h [hunk 5]
```diff
@@ -31,6 +32,47 @@ void smart_delete (SmartPointer<TYPE> &sp) DEAL_II_DEPRECATED;
...
+{
+public:
+
+  TimeUtilities()
+    :
+    status(true),
+    times()
+  {}
+
+  void sleep(unsigned int t);
```
Comment (luca-heltai) [inline include/utilities.h:54]: Add doc

File: include/utilities.h [hunk 6]
```diff
@@ -31,6 +32,47 @@ void smart_delete (SmartPointer<TYPE> &sp) DEAL_II_DEPRECATED;
...
+class TimeUtilities
+{
+public:
+
+  TimeUtilities()
+    :
+    status(true),
+    times()
+  {}
+
```
Comment (luca-heltai) [inline include/utilities.h:50]: Add doc

File: include/utilities.h [hunk 7]
```diff
@@ -31,6 +32,47 @@ void smart_delete (SmartPointer<TYPE> &sp) DEAL_II_DEPRECATED;
...
+  void sleep(unsigned int t);
+  void get_start_time();
+  void get_end_time();
```
Comment (luca-heltai) [inline include/utilities.h:64]: Add doc

File: include/utilities.h [hunk 8]
```diff
@@ -31,6 +32,47 @@ void smart_delete (SmartPointer<TYPE> &sp) DEAL_II_DEPRECATED;
...
+public:
+
+  TimeUtilities()
+    :
+    status(true),
+    times()
+  {}
+
+  void sleep(unsigned int t);
+  void get_start_time();
```
Comment (luca-heltai) [inline include/utilities.h:59]: Add doc

File: include/utilities.h [hunk 9]
```diff
@@ -79,6 +121,201 @@ bool copy_file(const std::string &files, const std::string &destination);
...
+ * (default = std::cout), the number of lines @p n_lines, and the width of the
+ * output.
+ *
+ * - get_shell_cols returns the number of coloumns of the shell.
+ * - get_shell_rows returns the number of rows of the shell.
+ * - get_current_line returns the current line the class is writing.
+ *
+ */
+template<typename Stream>
+class FilteredStream
```
Comment (luca-heltai) [inline include/utilities.h]: I'm not sure that `FilteredStream` is the right name. Maybe `OverWriteStream`?

File: include/utilities.h [hunk 10]
```diff
@@ -3,14 +3,15 @@
...
-#include <cxxabi.h>
 #include <deal.II/base/exceptions.h>
 #include <deal.II/fe/fe.h>
-
 #include <deal.II/base/std_cxx11/shared_ptr.h>
 
+#include <typeinfo>
+#include <cxxabi.h>
 #include <sstream>
+#include <sys/ioctl.h>    // to know the number of cols and rows of a shell
```
Comment (luca-heltai) [inline include/utilities.h:13]: Where do we need this?

Comment (ESeNonFossiIo) [inline include/utilities.h:13]: I would like the user chooses the length of each rows and that if the row length is too long ( greater than the shell length ) an assert fails.

File: include/utilities.h (hunk 9 above)
Comment (ESeNonFossiIo) [inline include/utilities.h]: ok. I change it

**Comment by ESeNonFossiIo** [comment]
When the output is seen in a shell ( runtime or using cat $file ) it seems to be ok.
Saving the output on a file and opening it with an editor or with cat -v $file shows all hidden chars... 

Solutions could be:
- rewind, seekp, and tellp -> do not work std::cout
- redirect stdcout to std::cout and file and viceversa -> I am not able to do it
- leave it as it is -> time saving option! :D

File: include/utilities.h [hunk 11]
```diff
@@ -79,6 +136,225 @@ bool copy_file(const std::string &files, const std::string &destination);
...
+ * This class uses @p n_lines lines of @p stream_out to show the output.
+ * Everytime it reaches the last line it comes back to the first line and
+ * rewrites the line.
+ *
+ * The constructor takes as argument a stream for the output @p stream_out
+ * (default = std::cout), the number of lines @p n_lines, and the width of the
+ * output.
+ *
+ */
+template<typename Stream>
```
Comment (luca-heltai) [inline include/utilities.h:152]: typename Stream=std::ostream

File: tests/fixed_lines_01.cc [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+int main ()
+{
+
+  TimeUtilities tu;
+
+  unsigned int  N = 3;
+  int           a = 100;
+  double        b = 3.14;
+
+  OverWriteStream<std::ostream> out(std::cout, N);
```
Comment (luca-heltai) [inline tests/fixed_lines_01.cc]: With the default value above, this should become

`OverWriteStream<> out(N);`

For this to work, the default value of `std::cout` has to be the second argument, not the first.  Change it in the constructor.

Comment (ESeNonFossiIo) [inline tests/fixed_lines_01.cc]: :+1:

File: include/utilities.h (hunk 11 above)
Comment (ESeNonFossiIo) [inline include/utilities.h:152]: It was in the forward declaration : 

``` cpp
template<typename Stream = std::ostream> class OverWriteStream;
```

**Pull request merged**